src/Passes/Source/RuleTransformationPass/RulePass.hpp

```diff
@@ -0,0 +1,245 @@
+    ///                    │                ┌───────────────────────────────┐  │  leave dead
+    ///                    │                │                               │  │    code?
+    ///                    │                │      Simplify phi nodes       │  │
+    ///                    │                │                               │  │
+    ///                    │                └───────────────────────────────┘  │
+    ///                    │                                │                  │
+    ///                    │                                │                  │
+    ///                    │                                ▼                  │
+    ///                    │                ┌───────────────────────────────┐  │
+    ///                    │                │                               │  │
+    ///                    │                │       Delete dead code        │  │
+    ///                    │                │                               │  │
+    ///                    │                └───────────────────────────────┘  │
+    ///                    │                                │                  │
+    ///                    │                                │                  │
+    ///                    │                                ▼                  │
+    ///                    │                ┌───────────────────────────────┐  │
+    ///                    │    fallback    │                               │  │
+    ///                    └───────────────▶│          Apply rules          │◀─┘
+    ///                                     │                               │
+    ///                                     └───────────────────────────────┘
+    ///
+    class RuleTransformationPass : public llvm::PassInfoMixin<RuleTransformationPass>
+    {
+      public:
+        using Replacements         = ReplacementRule::Replacements;
+        using Instruction          = llvm::Instruction;
+        using Rules                = std::vector<ReplacementRule>;
+        using Value                = llvm::Value;
+        using Builder              = ReplacementRule::Builder;
+        using AllocationManagerPtr = IAllocationManager::AllocationManagerPtr;
+        using Captures             = RuleSet::Captures;
+        using ConstantArguments    = std::unordered_map<std::string, llvm::ConstantInt*>;
+        using ILoggerPtr           = std::shared_ptr<ILogger>;
+
+        // Construction and destruction configuration.
+        //
+
+        /// Custom default constructor
+        RuleTransformationPass(RuleSet&& rule_set, RuleTransformationPassConfiguration const& config, Profile* profile);
+
+        /// Copy construction is banned.
+        RuleTransformationPass(RuleTransformationPass const&) = delete;
+
+        /// We allow move semantics.
+        RuleTransformationPass(RuleTransformationPass&&) = default;
+
+        /// Default destruction.
+        ~RuleTransformationPass() = default;
```

Comment:
(FYI) In some books (or articles or talks) I saw that the entities with `= default` they call 
"default**ed**" rather than 
"default".
Otherwise the expression "Default destructor" makes a confusing impression that there can also be a number of non-default destructors (same as default constructor and a number of non-default constructors). Or, when talking about the default constructor (a constructor that can be called without arguments) which also has `=default`, we have to use the word "default" twice in a row (which would sound confusing), whereas they call it "Default**ed** default constructor".

Reply:
I am a little bit puzzled about this comment. Do you have a reference where I can read about this topic?

Reply:
Not a good example but still... See [here](https://moodle.ufsc.br/pluginfile.php/2377667/mod_resource/content/0/Effective_Modern_C__.pdf), search for "defaulted" (Index starting with p.309).
[This](https://www.bogotobogo.com/cplusplus/files/CplusplusConcurrencyInAction_PracticalMultithreading.pdf) is a better example (search for "defaulted").

Reply:
Thanks, I will have a look at this on Monday.